File: praca-inzynierska-frontend/src/app/app/model/dto/PreferencesWeekDto.ts
import { PreferencesDayDto } from './PreferencesDayDto';
import { PreferencesWeekBackendDto } from './backend/PreferencesWeekBackendDto';

export class PreferencesWeekDto {

    constructor(
        public mondayPreferences: PreferencesDayDto,
        public tuesdayPreferences: PreferencesDayDto,
        public wednesdayPreferences: PreferencesDayDto,
        public thursdayPreferences: PreferencesDayDto,
        public fridayPreferences: PreferencesDayDto,
        public saturdayPreferences: PreferencesDayDto,
        public sundayPreferences: PreferencesDayDto
    ) {}

    static of(pw: PreferencesWeekBackendDto): PreferencesWeekDto {
        return new PreferencesWeekDto(
            PreferencesDayDto.of(pw.mondayPreferences),
            PreferencesDayDto.of(pw.tuesdayPreferences),
            PreferencesDayDto.of(pw.wednesdayPreferences),
            PreferencesDayDto.of(pw.thursdayPreferences),
            PreferencesDayDto.of(pw.fridayPreferences),
            PreferencesDayDto.of(pw.saturdayPreferences),
            PreferencesDayDto.of(pw.sundayPreferences)
        )
    }

    sortPreferences() {
        for (let i = 0; i < 7; i++) {
            this.getDay(i).sortPreferences()
        }
    }

    fillEmptyPeriodsWithDefaultPreference() {
        for (let i = 0; i < 7; i++) {
            this.getDay(i).fillEmptyPeriodsWithDefaultPreference()
        }
    }

    filterOutDefaultPreferences() {
        for (let i = 0; i < 7; i++) {
            this.getDay(i).filterOutDefaultPreferences()
        }
    }

    getDay(dayNumber: number): PreferencesDayDto {
        switch (dayNumber) {
            case 0:
                return this.mondayPreferences
            case 1:
                return this.tuesdayPreferences
            case 2:
                return this.wednesdayPreferences
            case 3:
                return this.thursdayPreferences
            case 4:
                return this.fridayPreferences
            case 5:
                return this.saturdayPreferences
            case 6:
                return this.sundayPreferences
            default:
                return undefined
        }
    }

    copy(): PreferencesWeekDto {
        return new PreferencesWeekDto(
            this.mondayPreferences.copy(),
            this.tuesdayPreferences.copy(),
            this.wednesdayPreferences.copy(),
            this.thursdayPreferences.copy(),
            this.fridayPreferences.copy(),
            this.saturdayPreferences.copy(),
            this.sundayPreferences.copy()
        )
    }

}

File: praca-inzynierska-backend/application/src/main/kotlin/jwer/backend/model/dto/PreferencesWeekDto.kt
package jwer.backend.model.dto

import jwer.backend.utils.TimePointArray
import jwer.backend.model.entity.PreferenceType
import jwer.backend.model.entity.PreferencesWeek
import jwer.backend.utils.iterator
import java.time.DayOfWeek
import java.time.LocalDate

data class PreferencesWeekDto(
        val mondayPreferences: PreferencesDayDto,
        val tuesdayPreferences: PreferencesDayDto,
        val wednesdayPreferences: PreferencesDayDto,
        val thursdayPreferences: PreferencesDayDto,
        val fridayPreferences: PreferencesDayDto,
        val saturdayPreferences: PreferencesDayDto,
        val sundayPreferences: PreferencesDayDto
) {
    constructor(pw: PreferencesWeek) : this(
            PreferencesDayDto(pw.mondayPreferences),
            PreferencesDayDto(pw.tuesdayPreferences),
            PreferencesDayDto(pw.wednesdayPreferences),
            PreferencesDayDto(pw.thursdayPreferences),
            PreferencesDayDto(pw.fridayPreferences),
            PreferencesDayDto(pw.saturdayPreferences),
            PreferencesDayDto(pw.sundayPreferences)
    )

    fun toPreferenceArray(firstDay: LocalDate, lastDay: LocalDate): Array<Int> {
        val array = TimePointArray<PreferenceType>(firstDay, lastDay, { it.toNumber() }, initialValue = PreferenceType.AVAILABLE.toNumber())
        for (date in firstDay..lastDay) {
            getPreferenceByDate(date).preferences.forEach {
                array[it.timePeriod.atDate(date)] = it.type
            }
        }
        return array.array
    }

    private fun getPreferenceByDate(date: LocalDate): PreferencesDayDto {
        return when (date.dayOfWeek) {
            DayOfWeek.MONDAY -> mondayPreferences
            DayOfWeek.TUESDAY -> tuesdayPreferences
            DayOfWeek.WEDNESDAY -> wednesdayPreferences
            DayOfWeek.THURSDAY -> thursdayPreferences
            DayOfWeek.FRIDAY -> fridayPreferences
            DayOfWeek.SATURDAY -> saturdayPreferences
            DayOfWeek.SUNDAY -> sundayPreferences
            null -> PreferencesDayDto()
        }
    }
}